src/LinearSolverTypes.f90

```diff
@@ -1615,228 +1662,216 @@ SUBROUTINE solveBiCGSTAB(solver)
-  TYPE(RealVectorType) :: u
-  INTEGER(SIK) :: k,m,n,it
-  TYPE(ParamType) :: pList
+  CALL vecPlist%clear()
+  CALL VectorResembleAlloc(Vy,thisLS%x,vecPlist)
+  CALL vecPlist%clear()
 
-  n=0
-  !Set parameter list for vector
-  CALL pList%add('VectorType -> n',solver%A%n)
-  n=solver%A%n
-  m=MIN(solver%nRestart,n)
-  CALL u%init(pList)
-  CALL pList%clear()
-  CALL solver%getResidual(u)
-  CALL PreCondType%apply(u)
-  CALL LNorm(u%b,2,beta)
-  solver%iters=0
+  SELECT TYPE(x => thisLS%X)
+  TYPE IS(RealVectorType)
+    CALL vecPlist%add('VectorType -> n',thisLS%A%n)
+#ifdef HAVE_MPI
+  TYPE IS(NativeDistributedVectorType)
+    CALL vecPlist%add('VectorType -> n',thisLS%A%n)
+    CALL vecPlist%add('VectorType -> MPI_Comm_ID',thisLS%MPIparallelEnv%comm)
+    CALL vecPlist%add('VectorType -> chunkSize',x%chunkSize)
+    CALL vecPlist%add('VectorType -> nlocal',SIZE(x%b))
+#endif
+  CLASS DEFAULT
+    CALL eLinearSolverType%raiseError('Incorrect call to '// &
+       modName//'::solveGMRES_partial'//' - Native solver does not support this vector type.')
+  ENDSELECT
 
-  IF(beta > EPSILON(0.0_SRK)) THEN
-    ALLOCATE(v(n,m+1))
-    ALLOCATE(R(m+1,m+1))
-    ALLOCATE(w(n))
-    ALLOCATE(c(m+1))
-    ALLOCATE(s(m+1))
-    ALLOCATE(g(m+1))
-    ALLOCATE(y(m+1))
-    v(:,:)=0._SRK
-    R(:,:)=0._SRK
-    w(:)=0._SRK
-    c(:)=0._SRK
-    s(:)=0._SRK
-    g(:)=0._SRK
-    y(:)=0._SRK
-    tol=solver%absConvTol*beta
-
-    v(:,1)=-u%b/beta
-    h=beta
-    phibar=beta
-#ifdef FUTILITY_DEBUG_MSG
-      WRITE(668,*) '         GMRES-LP',0,ABS(phibar)
+  ! Allocate Data storage arrays
+  SELECT TYPE(x => thisLS%X)
+  TYPE IS(RealVectorType)
+    ALLOCATE(RealVectorType :: V(thisLS%nRestart+1))
+#ifdef HAVE_MPI
+  TYPE IS(NativeDistributedVectorType)
+    ALLOCATE(NativeDistributedVectorType :: V(thisLS%nRestart+1))
 #endif
-    !Iterate on solution
-    DO it=1,m
-      CALL BLAS_matvec(THISMATRIX=solver%A,X=v(:,it),BETA=0.0_SRK,Y=w)
-      u%b=w
-      CALL solver%PreCondType%apply(u)
-      w=u%b
-      h=BLAS_dot(n,w,1,v(:,1),1)
-      w=w-h*v(:,1)
-      t=h
-      DO k=2,it
-        h=BLAS_dot(n,w,1,v(:,k),1)
-        w=w-h*v(:,k)
-        R(k-1,it)=c(k-1)*t+s(k-1)*h
-        t=c(k-1)*h-s(k-1)*t
-      ENDDO
-      CALL LNorm(w,2,h)
-      !WRITE(*,*) "h = ", h
-      IF(h > 0.0_SRK) THEN
-        v(:,it+1)=w/h
-      ELSE
-        v(:,it+1)=0.0_SRK*w
-      ENDIF
-      !Set up next Given's rotation
-      IF(t >= 0.0_SRK) THEN
-        temp=SQRT(t*t+h*h)
-      ELSE
-        temp=-SQRT(t*t+h*h)
-      ENDIF
-      c(it)=t/temp
-      s(it)=h/temp
-      R(it,it)=temp
-      g(it)=c(it)*phibar
-      phibar=-s(it)*phibar
-#ifdef FUTILITY_DEBUG_MSG
-      WRITE(668,*) '         GMRES-LP',it,ABS(phibar)
+  ENDSELECT
+
+  ALLOCATE(h(thisLS%nRestart))
+  ALLOCATE(R(thisLS%nRestart+1,thisLS%nRestart+1))
+  ALLOCATE(givens_cos(thisLS%nRestart+1))
+  ALLOCATE(givens_sin(thisLS%nRestart+1))
+  ALLOCATE(g(thisLS%nRestart+1))
```

Comment:
These 4 allocations didn't have `+1` in them last time.  Why that change?

Reply:
This is wrapped up in the whole "older version of the solver" thing, and it's related to how the orthogonalization and given's rotation constants are stored. What it comes down to is that in the current version, you loop in such a way that you generate these constants for the basis vector i+1 (having started with the constants for basis vector 1 already initialized).
Before, you would work on basis vector i+1 while generating the constants for basis vector i.

I believe there is an argument to be made that this should be more efficient in terms of achieving the linear solver tolerance, since it effectively lets you use an additional basis vector which is wasted in the previous case. This is part of the rationale for changing the solver to this version.

Reply:
Alright, cool.  Just making sure this was an intended change.